src/shared/voice/subscription.ts

```diff
@@ -0,0 +1,152 @@
+						// Probably moved voice channel
+					} catch {
+						this.voiceConnection.destroy();
+						// Probably removed from voice channel
+					}
+				} else if (this.voiceConnection.rejoinAttempts < 5) {
+					/*
+						The disconnect in this case is recoverable, and we also have <5 repeated attempts so we will reconnect.
+					*/
+					await wait((this.voiceConnection.rejoinAttempts + 1) * 5_000);
+					this.voiceConnection.rejoin();
+				} else {
+					/*
+						The disconnect in this case may be recoverable, but we have no more remaining attempts - destroy.
+					*/
+					this.voiceConnection.destroy();
+				}
+			} else if (newState.status === VoiceConnectionStatus.Destroyed) {
+				/*
+					Once destroyed, stop the subscription
+				*/
+				this.stop();
+			} else if (
+				!this.readyLock &&
+				(newState.status === VoiceConnectionStatus.Connecting || newState.status === VoiceConnectionStatus.Signalling)
+			) {
+				/*
+					In the Signalling or Connecting states, we set a 20 second time limit for the connection to become ready
+					before destroying the voice connection. This stops the voice connection permanently existing in one of these
+					states.
+				*/
+				this.readyLock = true;
+				try {
+					await entersState(this.voiceConnection, VoiceConnectionStatus.Ready, 20_000);
+				} catch {
+					if (this.voiceConnection.state.status !== VoiceConnectionStatus.Destroyed) this.voiceConnection.destroy();
+				} finally {
+					this.readyLock = false;
+				}
+			}
+		});
+
+		// Configure audio player
+		this.audioPlayer.on('stateChange', (oldState, newState) => {
+			if (newState.status === AudioPlayerStatus.Idle && oldState.status !== AudioPlayerStatus.Idle) {
+				// If the Idle state is entered from a non-Idle state, it means that an audio resource has finished playing.
+				// The queue is then processed to start playing the next track, if one is available.
+				(oldState.resource as AudioResource<Track>).metadata.onFinish((oldState.resource as AudioResource<Track>).metadata.title);
+				this.processQueue();
+			} else if (newState.status === AudioPlayerStatus.Playing) {
+				// If the Playing state has been entered, then a new track has started playback.
+				(newState.resource as AudioResource<Track>).metadata.onStart((newState.resource as AudioResource<Track>).metadata.title);
+			}
+		});
+
+		this.audioPlayer.on('error', (error) => (error.resource as AudioResource<Track>).metadata.onError((error.resource as AudioResource<Track>).metadata.title, error));
+
+		voiceConnection.subscribe(this.audioPlayer);
+	}
+
+	/**
+	 * Adds a new Track to the queue.
+	 *
+	 * @param track The track to add to the queue
+	 */
+	public enqueue(track: Track) {
+		this.queue.push(track);
+		this.processQueue();
+	}
+
+	/**
+	 * Stops audio playback and empties the queue
+	 */
+	public stop() {
+		this.queueLock = true;
+		this.queue = [];
+		this.audioPlayer.stop(true);
```

Comment:
Feel like should unlock queueLock here

Reply:
When using playlists, if the queue is unlocked right after a clear, and the playlist is still loading, it will load the rest into the queue. That's why it's not unlocked immediately.

Reply:
It now unlocks queuelock here.